lib/common/client/concerns/mhv_locked_session_client.rb

```diff
@@ -0,0 +1,121 @@
+# frozen_string_literal: true
+
+module Common
+  module Client
+    module Concerns
+      ##
+      # Module mixin for overriding session logic when making session-based client connections that
+      # should lock during session creation, to prevent threads from making simultaneous
+      # authentication API calls.
+      #
+      # All refrences to "session" in this module refer to the upstream MHV session.
+      #
+      # @see MedicalRecords::Client
+      #
+      # @!attribute [r] session
+      #   @return [Hash] a hash containing session information
+      #
+      module MhvLockedSessionClient
+        extend ActiveSupport::Concern
+        include SentryLogging
+
+        LOCK_RETRY_DELAY = 1 # Number of seconds to wait between attempts to acquire a session lock
+        RETRY_ATTEMPTS = 10 # How many times to attempt to acquire a session lock
+
+        attr_reader :session
+
+        ##
+        # @param session [Hash] a hash containing a key with which the session will be found or built
+        #
+        def initialize(session:)
+          refresh_session(session)
+        end
+
+        ##
+        # Ensure the upstream MHV session is not expired or incomplete.
+        #
+        # @return [MhvFhirSessionClient] instance of `self`
+        #
+        def authenticate
+          raise 'ICN is required for session creation' unless user_key
+
+          iteration = 0
+
+          # Loop unless a complete, valid MHV session exists, or until max_iterations is reached
+          while invalid?(session) && iteration < RETRY_ATTEMPTS
```

Comment:
I'm weary of `while`. Would it make sense to add an escape hatch if the elapsed time exceeded delay * attempts?

Reply:
I misunderstood how while and && interface. I think this is sufficient as long as interation += 1 is always hit.